File: domain/src/main/java/com/beat/domain/cast/domain/Cast.java
package com.beat.domain.cast.domain;

import lombok.AccessLevel;
import lombok.Builder;
import lombok.Getter;
import lombok.NoArgsConstructor;

@Getter
@NoArgsConstructor(access = AccessLevel.PROTECTED)
public class Cast {

	private Long id;

	private String castName;

	private String castRole;

	private String castPhoto;

	private Long performanceId;

	@Builder(access = AccessLevel.PRIVATE)
	private Cast(Long id, String castName, String castRole, String castPhoto, Long performanceId) {
		this.id = id;
		this.castName = castName;
		this.castRole = castRole;
		this.castPhoto = castPhoto;
		this.performanceId = performanceId;
	}

	public static Cast create(String castName, String castRole, String castPhoto, Long performanceId) {
		return Cast.builder()
			.castName(castName)
			.castRole(castRole)
			.castPhoto(castPhoto)
			.performanceId(performanceId)
			.build();
	}

	public static Cast rehydrate(Long id, String castName, String castRole, String castPhoto, Long performanceId) {
		return Cast.builder()
			.id(id)
			.castName(castName)
			.castRole(castRole)
			.castPhoto(castPhoto)
			.performanceId(performanceId)
			.build();
	}

	public void update(String castName, String castRole, String castPhoto) {
		this.castName = castName;
		this.castRole = castRole;
		this.castPhoto = castPhoto;
	}
}

File: domain/src/main/kotlin/com/beat/domain/cast/domain/Cast.kt
package com.beat.domain.cast.domain

import com.beat.domain.performance.domain.Performance

@ConsistentCopyVisibility
data class Cast private constructor(
    private val castId: Id?,
    val castName: String,
    val castRole: String,
    val castPhoto: String,
    private val linkedPerformanceId: Performance.Id,
) {
    fun getId(): Long? = castId?.value

    fun getPerformanceId(): Long = linkedPerformanceId.value

    fun update(castName: String, castRole: String, castPhoto: String): Cast = copy(
        castName = castName,
        castRole = castRole,
        castPhoto = castPhoto
    )

    @JvmInline
    value class Id private constructor(val value: Long) {
        companion object {
            @JvmStatic
            fun from(value: Long): Id = Id(value)

            @JvmStatic
            fun fromNullable(value: Long?): Id? = value?.let(::from)
        }
    }

    companion object {
        @JvmStatic
        fun create(castName: String, castRole: String, castPhoto: String, performanceId: Long): Cast = Cast(
            castId = null,
            castName = castName,
            castRole = castRole,
            castPhoto = castPhoto,
            linkedPerformanceId = Performance.Id.from(performanceId)
        )

        @JvmStatic
        fun rehydrate(id: Long?, castName: String, castRole: String, castPhoto: String, performanceId: Long): Cast =
            Cast(
                castId = Id.fromNullable(id),
                castName = castName,
                castRole = castRole,
                castPhoto = castPhoto,
                linkedPerformanceId = Performance.Id.from(performanceId)
            )
    }
}
